test/wasm2c/tail-calls.txt

```diff
@@ -0,0 +1,953 @@
+    }
+    va_end(args);
+  }
+  
+  return NULL;
+}
+
+void w2c_test_tailcaller_0(w2c_test* instance) {
+  FUNC_PROLOGUE;
+  u32 var_i0, var_i2;
+  f32 var_f1;
+  var_i0 = 1u;
+  var_f1 = 2;
+  var_i2 = 0u;
+  static_assert(sizeof(struct wasm_multi_if) <= 1024);
+  CHECK_CALL_INDIRECT(instance->w2c_tab, w2c_test_i32_f32, var_i2);
+  if (!instance->w2c_tab.data[var_i2].func_tailcallee.fn) {
+    CALL_INDIRECT(instance->w2c_tab, void (*)(void*, u32, f32), w2c_test_i32_f32, var_i2, instance->w2c_tab.data[var_i2].module_instance, var_i0, var_f1);
+  } else {
+    void *instance_ptr_storage;
+    void **instance_ptr = &instance_ptr_storage;
+    char tail_call_stack[1024];
+    wasm_rt_tailcallee_t next_storage;
+    wasm_rt_tailcallee_t *next = &next_storage;
+    {
+      struct wasm_multi_if *tmp = (struct wasm_multi_if *)tail_call_stack;
+      tmp->i0 = var_i0;
+      tmp->f1 = var_f1;
+    }
+    next->fn = instance->w2c_tab.data[var_i2].func_tailcallee.fn;
+    *instance_ptr = instance->w2c_tab.data[var_i2].module_instance;
+    while (next->fn) { next->fn(instance_ptr, tail_call_stack, next); }
+  }
+  goto var_Bfunc;
+  var_Bfunc:;
+  FUNC_EPILOGUE;
+}
+
+void wasm_tailcall_w2c_test_tailcaller_0(void **instance_ptr, void *tail_call_stack, wasm_rt_tailcallee_t *next) {
+  w2c_test* instance = *instance_ptr;
+  u32 var_i0, var_i2;
+  f32 var_f1;
+  var_i0 = 1u;
+  var_f1 = 2;
+  var_i2 = 0u;
+  static_assert(sizeof(struct wasm_multi_if) <= 1024);
+  CHECK_CALL_INDIRECT(instance->w2c_tab, w2c_test_i32_f32, var_i2);
+  if (!instance->w2c_tab.data[var_i2].func_tailcallee.fn) {
+    CALL_INDIRECT(instance->w2c_tab, void (*)(void*, u32, f32), w2c_test_i32_f32, var_i2, instance->w2c_tab.data[var_i2].module_instance, var_i0, var_f1);
+    next->fn = NULL;
+  } else {
+    {
+      struct wasm_multi_if *tmp = (struct wasm_multi_if *)tail_call_stack;
+      tmp->i0 = var_i0;
+      tmp->f1 = var_f1;
+    }
+    next->fn = instance->w2c_tab.data[var_i2].func_tailcallee.fn;
+    *instance_ptr = instance->w2c_tab.data[var_i2].module_instance;
+  }
+  return;
+  var_Bfunc:;
+  next->fn = NULL;
+}
+
+struct wasm_multi_id w2c_test_infiniteloop(w2c_test* instance, u32 var_p0, f64 var_p1) {
+  FUNC_PROLOGUE;
+  u32 var_i0;
+  f64 var_d1;
+  var_i0 = var_p0;
+  var_d1 = var_p1;
+  static_assert(sizeof(struct wasm_multi_id) <= 1024);
+  {
+    void *instance_ptr_storage;
+    void **instance_ptr = &instance_ptr_storage;
+    char tail_call_stack[1024];
+    wasm_rt_tailcallee_t next_storage;
+    wasm_rt_tailcallee_t *next = &next_storage;
+    {
+      struct wasm_multi_id *tmp = (struct wasm_multi_id *)tail_call_stack;
+      tmp->i0 = var_i0;
+      tmp->d1 = var_d1;
+    }
+    next->fn = wasm_tailcall_w2c_test_infiniteloop;
+    while (next->fn) { next->fn(instance_ptr, tail_call_stack, next); }
+    {
+      struct wasm_multi_id *tmp = tail_call_stack;
+      var_i0 = tmp->i0;
+      var_d1 = tmp->d1;
+    }
+  }
+  goto var_Bfunc;
+  var_Bfunc:;
```

Comment:
It looks like wasm2c has done this since the beginning (https://github.com/WebAssembly/wabt/blob/main/src/c-writer.cc#L1111); I think the reason is to make the code correct even when the label occurs at the very end of a function. (A label must precede a statement but the null statement is okay.)

Reply:
ah gotcha! :+1: